experiments/run_single_experiment.py

```diff
@@ -25,56 +26,138 @@ def _main(cfg: DictConfig) -> None:
+        cfg.env, scene_spec=cfg.scene_spec, seed=cfg.seed
+    )
+    assert isinstance(train_env, gym.Env)
+    if cfg.record_train_videos:
+        train_env = gym.wrappers.RecordVideo(
+            train_env, str(Path(cfg.video_dir) / "train")
+        )
+    train_env.action_space.seed(cfg.seed)
+
+    # Create eval environment, which will be reset all the time.
+    eval_seed = cfg.seed + cfg.eval_seed_offset
+    eval_env = hydra.utils.instantiate(
+        cfg.env, scene_spec=cfg.scene_spec, seed=eval_seed
+    )
+    assert isinstance(eval_env, gym.Env)
+    if cfg.record_eval_videos:
+        eval_env = gym.wrappers.RecordVideo(eval_env, str(Path(cfg.video_dir) / "eval"))
+    eval_env.action_space.seed(eval_seed)
+
+    # Create two copies of the approach. The eval approach will load model files
+    # from the train approach to do evaluation without losing track of state in
+    # the training approach.
+    train_approach = hydra.utils.instantiate(
         cfg.approach,
-        action_space=env.action_space,
+        scene_spec=cfg.scene_spec,
+        action_space=train_env.action_space,
         seed=cfg.seed,
     )
-    assert isinstance(approach, BaseApproach)
-    approach.train()
-
-    # Run a certain number of time steps and log metrics along the way.
-    metrics: list[dict[str, float]] = []
-    obs, info = env.reset()
-    assert "user_allows_explore" in info, (
-        "Environments are required to report at reset whether the robot should "
-        "explore or not. The user decides."
+    assert isinstance(train_approach, BaseApproach)
+    train_approach.train()
+    eval_approach = hydra.utils.instantiate(
+        cfg.approach,
+        scene_spec=cfg.scene_spec,
+        action_space=eval_env.action_space,
+        seed=eval_seed,
     )
-    approach.reset(obs, info)
-    for t in range(cfg.max_environment_steps):
+    assert isinstance(eval_approach, BaseApproach)
+    eval_approach.eval()
+
+    # Log training and eval metrics separately.
+    train_metrics: list[dict[str, float]] = []
+    eval_metrics: list[dict[str, float]] = []
+
+    # Reset the training environment, one time only.
+    obs, info = train_env.reset()
+    # Reset the training approach, one time only.
+    train_approach.reset(obs, info)
+    # Main training and eval loop.
+    for t in range(cfg.max_environment_steps + 1):
+        # Check if it's time to eval.
+        if t % cfg.eval_frequency == 0:
+            # Save the models from the training approach and load them into the
+            # eval approach.
+            step_model_dir = model_dir / str(t)
+            step_model_dir.mkdir(exist_ok=True)
+            train_approach.save(step_model_dir)
+            eval_approach.load(step_model_dir)
+            # Run evaluation.
+            step_eval_metrics = _evaluate_approach(eval_approach, eval_env, cfg, t)
+            eval_metrics.append(step_eval_metrics)
+        # Eval on the last time step but don't train anymore.
```

Comment:
Consider adding an `assert cfg.max_environment_steps % cfg.eval_frequency == 0` when loading the config to make sure of this